lib-src/libcs/atoh.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" atoh.3,v
.\" Revision 1.1.1.1  1996/11/22 19:19:28  rvb
.\"  almost done
.\"
.\" Revision 1.1  1996/11/22 19:19:27  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  18:59:54  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:57:42  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:52:36  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" atoh.3,v
.\" Revision 1.1.1.1  1996/11/22 19:19:28  rvb
.\"  almost done
.\"
.\" Revision 1.1  1996/11/22 19:19:27  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  18:59:54  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:57:42  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:52:36  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:52:36  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.3  90/12/12  15:45:56  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH ATOH 3 12/5/79
.CM 1
.SH "NAME"
atoh, atoo \- convert ASCII to hexadecimal or octal
.SH "SYNOPSIS"
.B
unsigned int atoh (ap)
.br
.B
char *ap;
.sp
.B
unsigned int atoo (ap)
.br
.B
char *ap;
.SH "DESCRIPTION"
.I
Atoh
converts the string pointed to by
.I
ap,
representing an hexadecimal number, to an unsigned integer.
The string may contain leading blanks and tabs, and a string of
hexadecimal digits.
Conversion stops at the first character which is
not a valid hexadecimal digit.
The digits themselves include the characters "0" through "9",
"a" through "f", and "A" through "F".
.sp
.I
Atoo
converts the string pointed to by
.I
ap,
representing an octal number, to an unsigned integer.
The string may contain leading blanks and tabs, and a string of
octal digits.
Conversion stops at the first character which is
not a valid octal digit.
.SH "LIBRARY NOTES"
You must link agains /usr/local/lib/libcs.a to define these routine.
.SH "SEE ALSO"
.IR atof (3), 
.IR atoi (3), 
.IR atol (3), 
.SH "BUGS"
There is no provision for overflow.
Unary + is not accepted.
(Unary
- is not accepted, due to the interpretation of the numbers as unsigned.)


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/atoh.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: /afs/cs.cmu.edu/project/coda-braam/src/coda-4.0.1/RCSLINK/./lib-src/libcs/atoh.c,v 1.1 1996/11/22 19:19:45 braam Exp $";
#endif /*_BLURB_*/

/*
 *  atoh  --  convert ascii to hexidecimal
 *
 *  Usage:  i = atoh (string);
 *	unsigned int i;
 *	char *string;
 *
 *  Atoo converts the value contained in "string" into an
 *  unsigned integer, assuming that the value represents
 *  a hexidecimal number.
 */

#include "libcs.h"

unsigned int atoh(ap)
const char *ap;
{
	register const char *p;
	register unsigned int n;
	register int digit, lcase = 0;

	p = ap;
	n = 0;
	while(*p == ' ' || *p == '	')
		p++;

	if(*p == '0' && ((*(p+1) == 'x') || (*(p+1) == 'X')))
		p+=2;

	while ((digit = (*p >= '0' && *p <= '9')) ||
		(lcase = (*p >= 'a' && *p <= 'f')) ||
		(*p >= 'A' && *p <= 'F')) {
		n *= 16;
		if (digit)	n += *p++ - '0';
		else if (lcase)	n += 10 + (*p++ - 'a');
		else		n += 10 + (*p++ - 'A');
	}
	return(n);
}
